tensorflow/core/kernels/tensor_forest/resources.cc

```diff
@@ -68,4 +71,180 @@ void TensorForestTreeResource::Reset() {
+          label,
+          candidate.left_split_stats().sum().value(label) + incoming_weight);
+
+      float new_weight = old_left_weight + incoming_weight;
+
+      candidate.mutable_left_split_stats()->mutable_sum_of_square()->set_value(
+          0, candidate.left_split_stats().sum_of_square().value(0) -
+                 old_left_weight * old_left_weight + new_weight * new_weight);
+
+      candidate.mutable_left_leaf_stats()->set_weight_sum(
+          incoming_weight + candidate.left_leaf_stats().weight_sum());
+
+      candidate.mutable_left_leaf_stats()->mutable_counts_or_sums()->set_value(
+          label, new_weight);
+
+    } else {
+      float old_right_weight = old_total_weight - old_left_weight;
+
+      candidate.mutable_right_split_stats()->mutable_sum()->set_value(
+          label,
+          candidate.right_split_stats().sum().value(label) + incoming_weight);
+
+      float new_weight = old_right_weight + incoming_weight;
+
+      candidate.mutable_right_split_stats()->mutable_sum_of_square()->set_value(
+          0, candidate.right_split_stats().sum_of_square().value(0) -
+                 old_right_weight * old_right_weight + new_weight * new_weight);
+    }
+  }
+};
+
+const bool TensorForestFertileStatsResource::BestSplitFromSlot(
+    const tensor_forest::FertileSlot& slot,
+    tensor_forest::SplitCandidate* best) const {
+  float min_score = FLT_MAX;
+  for (int i = 0; i < slot.candidates_size(); i++) {
+    auto candidate = slot->candidates(i);
+    // for (auto candidate : slot.candidates()) {
+    float left_gini =
+        1 - candidate.left_split_stats().sum_of_square().value(0) /
+                (candidate.left_split_stats().sum().value(0) *
+                 candidate.left_split_stats().sum().value(0));
+    float right_gini =
+        1 - candidate.right_split_stats().sum_of_square().value(0) /
+                (candidate.right_split_stats().sum().value(0) *
+                 candidate.right_split_stats().sum().value(0));
+
+    if (candidate.left_leaf_stats().weight_sum() > 0 &&
+        slot.leaf_stats().weight_sum() -
+                candidate.left_leaf_stats().weight_sum() >
+            0 &&
+        left_gini + right_gini < min_score) {
+      min_score = left_gini + right_gini;
+      best = &candidate;
+    }
+  }
+  return min_score != FLT_MAX;
+}
+
+const bool TensorForestFertileStatsResource::AddSplitToSlot(
+    const int32 node_id, const int32 feature_id, const float threshold,
+    const int32 example_id, const int32 num_targets,
+    const TTypes<float>::ConstMatrix* dense_feature,
+    const TTypes<float>::ConstMatrix* labels) {
+  auto slot = fertile_stats_->node_to_slot().at(node_id);
+  slot.mutable_leaf_stats()->set_weight_sum(slot.leaf_stats().weight_sum() + 1);
+  auto candidate = slot.add_candidates();
+  auto split = candidate->mutable_split();
+  split->set_threshold(threshold);
+  split->set_feature_id(feature_id);
+  float incoming_weight = 1.0;
```

Comment:
it looks like it is just a cont, move it to the top of the file, make it a real const